File: 11/haskell/main.hs
import Data.Array
import Data.Char
import Data.List
import Data.List.Split
import Data.Maybe
import System.IO

-- Uses haskell arrays because we're performing a bunch of index operations
type Position = (Int, Int)

type Grid = Array (Int, Int) Int

buildArray :: String -> Grid
buildArray value =
  array
    ((1, 1), (w, h))
    [ ((x, y), v)
      | (y, vs) <- zip [1 ..] vss,
        (x, v) <- zip [1 ..] vs
    ]
  where
    h = length vss
    w = length $ head vss
    vss = map (map digitToInt) $ lines value

prettyPrintGrid :: Grid -> String
prettyPrintGrid grid = foldr (\(x, y) acc -> (if y == 1 then "\n" else "") ++ show (grid ! (x, y)) ++ " " ++ acc) [] (indices grid)

validPosition :: ((Int, Int), (Int, Int)) -> Position -> Bool
validPosition ((minX, minY), (maxX, maxY)) (x, y) =
  x <= maxX && x >= minX && y <= maxY && y >= minY

getNeighbors :: ((Int, Int), (Int, Int)) -> Position -> [Position]
getNeighbors bounds (x, y) =
  filter
    (validPosition bounds)
    [(x + dx, y + dy) | dy <- [-1, 0, 1], dx <- [-1, 0, 1]]

incrementPositions :: Grid -> [Position] -> Grid
incrementPositions grid positions = grid // [(pos, grid ! pos + 1) | pos <- positions]

zeroPositions :: Grid -> [Position] -> Grid
zeroPositions grid positions = grid // [(pos, 0) | pos <- positions]

getFlashes :: Grid -> [Position]
getFlashes grid = foldl' (\acc x -> if grid ! x > 9 then x : acc else acc) [] $ indices grid

applyFlashes :: Grid -> [Position] -> (Grid, [Position])
applyFlashes grid flashed =
  if null flashes
    then (grid, flashed)
    else applyFlashes (foldl (\acc x -> incrementPositions acc [x]) grid (flashes >>= getNeighbors (bounds grid))) (flashes ++ flashed)
  where
    flashes = getFlashes grid \\ flashed

runGeneration :: Grid -> (Grid, Int)
runGeneration grid = (finalGrid, length flashes)
  where
    incremented = incrementPositions grid (indices grid)
    (flashed, flashes) = applyFlashes incremented []
    finalGrid = zeroPositions flashed flashes

runGenerations :: Grid -> Int -> Int
runGenerations _ 0 = 0
runGenerations grid i = flashes + runGenerations newGrid (i - 1)
  where
    (newGrid, flashes) = runGeneration grid

firstSync :: Grid -> Int -> Int
firstSync grid gen = if flashes == 100 then gen else firstSync newGrid gen + 1
  where
    (newGrid, flashes) = runGeneration grid

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (runGenerations (buildArray contents) 100)
  print (firstSync (buildArray contents) 1)
  hClose handle

File: 11/python/main.py

def solution(lines):
    grid = [[int(y) for y in x.strip()] for x in lines]
    for row in grid:
        row.append(-float('infinity'))
        row.insert(0, -float('infinity'))
    grid.append([-float('infinity') for _ in grid[0]])
    grid.insert(0, [-float('infinity') for _ in grid[0]])
    flashes = 0
    gen = 0
    while True:
        gen += 1
        print(gen)
        # for row in grid:
        #     print(''.join([str(x) for x in row]))
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                grid[i][j] += 1
        has_flashed = True
        flashed = []
        while has_flashed:
            
            has_flashed = False
            for i in range(len(grid)):
                for j in range(len(grid[i])):
                    
                    if grid[i][j] > 9 and not (i,j) in flashed:
                        
                        flashes += 1
                        flashed.append((i,j))
                        grid[i-1][j-1] += 1
                        grid[i-1][j] += 1
                        grid[i-1][j+1] += 1
                        grid[i][j-1] += 1
                        grid[i][j] += 1
                        grid[i][j+1] += 1
                        grid[i+1][j-1] += 1
                        grid[i+1][j] += 1
                        grid[i+1][j+1] += 1
                        has_flashed = True
                if has_flashed:
                    break
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if (grid[i][j] > 9):

                    grid[i][j] = 0
        
        if (len(flashed) == 100):
            print(gen)
            return
    print(flashes)
    
with open('input.txt') as f:
    lines = f.readlines()
    print(solution(lines))